keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTS.java

```diff
@@ -0,0 +1,546 @@
+                            u,
+                            v,
+                            j - 1,
+                            epsilon,
+                            random
+                    );
+
+                    positionForward = rightTree.positionForward;
+                    gradientForward = rightTree.gradientForward;
+                    momentumForward = rightTree.momentumForward;
+                    logOfMasterPAtPositionForward = rightTree.logOfMasterPAtPositionForward;
+                }
+
+                double acceptThetaPrimeProbability = (double) rightTree.nPrime / (leftTree.nPrime + rightTree.nPrime);
+
+                if (withProbability(acceptThetaPrimeProbability, random)) {
+                    thetaPrime = rightTree.thetaPrime;
+                    gradientAtThetaPrime = rightTree.gradientAtThetaPrime;
+                    logOfMasterPAtThetaPrime = rightTree.logOfMasterPAtThetaPrime;
+                    sampleAtThetaPrime = rightTree.sampleAtThetaPrime;
+                }
+
+                sPrime = rightTree.sPrime * isNotUTurning(
+                        positionForward,
+                        positionBackward,
+                        momentumForward,
+                        momentumBackward
+                );
+
+                nPrime = nPrime + rightTree.nPrime;
+            }
+
+            return new BuiltTree(
+                    positionBackward,
+                    gradientBackward,
+                    momentumBackward,
+                    logOfMasterPAtPositionBackward,
+                    positionForward,
+                    gradientForward,
+                    momentumForward,
+                    logOfMasterPAtPositionForward,
+                    thetaPrime,
+                    gradientAtThetaPrime,
+                    logOfMasterPAtThetaPrime,
+                    sampleAtThetaPrime,
+                    nPrime,
+                    sPrime
+            );
+        }
+
+    }
+
+    private static double getLogProb(List<Vertex<?>> probabilisticVertices) {
+        double sum = 0.0;
+        for (Vertex<?> vertex : probabilisticVertices) {
+            sum += vertex.logProbAtValue();
+        }
+        return sum;
+    }
+
+    private static boolean withProbability(double probability, Random random) {
+        return random.nextDouble() < probability;
+    }
+
+    private static int isNotUTurning(Map<String, Double> positionForward,
+                                     Map<String, Double> positionBack,
+                                     Map<String, Double> momentumForward,
+                                     Map<String, Double> momentumBack) {
+        double forward = 0.0;
+        double backward = 0.0;
+
+        for (Map.Entry<String, Double> forwardEntry : positionForward.entrySet()) {
+
+            final String id = forwardEntry.getKey();
+            final double forwardMinusBack = forwardEntry.getValue() - positionBack.get(id);
```

Comment:
nitpick: can we call this `forwardMinusBackward`? I got confused on what it was

Reply:
I struggled to suss this out, maybe it's overkill but could we do:

            `final double currentPositionForward = forwardEntry.getValue();`
            `final double forwardMinusBack = currentPositionForward - positionBack.get(id);`